store/src/java/com/zimbra/cs/service/formatter/LdifFormatter.java

```diff
@@ -0,0 +1,192 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.mail.MessagingException;
+import javax.mail.Part;
+import javax.servlet.ServletException;
+
+import org.apache.commons.codec.binary.Base64;
+
+import com.zimbra.common.localconfig.LC;
+import com.zimbra.common.mailbox.ContactConstants;
+import com.zimbra.common.mime.MimeConstants;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.util.HttpUtil;
+import com.zimbra.cs.mailbox.MailItem;
+import com.zimbra.cs.service.UserServletContext;
+import com.zimbra.cs.service.UserServletException;
+import com.zimbra.cs.service.formatter.FormatterFactory.FormatType;
+import com.zimbra.cs.service.util.UserServletUtil;
+import com.zimbra.cs.servlet.util.JettyUtil;
+
+public class LdifFormatter extends Formatter {
+
+    private static final String NEW_LINE = "\r\n";
+
+    @Override
+    public FormatType getType() {
+        return FormatType.LDIF;
+    }
+
+    @Override
+    public void formatCallback(UserServletContext context) throws UserServletException,
+        ServiceException, IOException, ServletException, MessagingException {
+        // Disable the jetty timeout
+        disableJettyTimeout(context);
+        Iterator<? extends MailItem> contacts = null;
+        StringBuilder sb = new StringBuilder();
+        sb.append("version: 1"); // ldif version
+        sb.append(NEW_LINE);
+        try {
+            contacts = getMailItems(context, -1, -1, Integer.MAX_VALUE);
+            ArrayList<Map<String, String>> allContacts = new ArrayList<Map<String, String>>();
+            HashSet<String> fields = new HashSet<String>();
+            UserServletUtil.populateContactFields(contacts, context.targetMailbox,
+                context.opContext, allContacts, fields);
+            ArrayList<String> allFields = new ArrayList<String>();
+            allFields.addAll(fields);
+            Collections.sort(allFields);
+            for (Map<String, String> contactMap : allContacts) {
+                toLDIFContact(allFields, contactMap, sb);
+                sb.append(NEW_LINE);
+            }
+        } finally {
+            if (contacts instanceof QueryResultIterator)
+                ((QueryResultIterator) contacts).finished();
+        }
+        String filename = context.itemPath;
+        if (filename == null || filename.length() == 0)
+            filename = "contacts";
+        if (!filename.toLowerCase().endsWith(".ldif")) {
+            filename = filename + ".ldif";
+        }
+        String cd = HttpUtil.createContentDisposition(context.req, Part.ATTACHMENT, filename);
+        context.resp.addHeader("Content-Disposition", cd);
+        context.resp.setCharacterEncoding(context.getCharset().name());
+        context.resp.setContentType(MimeConstants.CT_TEXT_LDIF);
+        context.resp.getWriter().print(sb.toString());
+    }
+
+    public void toLDIFContact(List<String> fields, Map<String, String> contact, StringBuilder sb) {
```

Comment:
Actually using an LDIF writer that is specifically designed to create LDIF (https://docs.ldap.com/ldap-sdk/docs/javadoc/com/unboundid/ldif/LDIFWriter.html) would be a good idea. As I mentioned, the implementation we have created doesn't seem to cover the requirement that base64 encoded lines be no more than 76 characters long (https://tools.ietf.org/html/rfc2045 as hinted at in https://tools.ietf.org/html/rfc2849).